src/openlcb/BroadcastTimeDefs.cxx

```diff
@@ -137,49 +193,113 @@ std::string BroadcastTimeDefs::date_to_string(int year, int month, int day)
 {
-    struct tm tm;
-    if (strptime(stime.c_str(), "%R", &tm) == nullptr)
+    // Note: The obvious way to simplify this method would be to make use of
+    //       strptime(). However, avoiding strptime() saves nearly 2K of code
+    //       space (ARMv7m, -Os).
+
+    bool result = true;
+    int h_value = 0;
+    int m_value = 0;
+
+    // Find the colon separator.
+    size_t colon = stime.find(':');
+    if (colon == std::string::npos)
     {
-        return false;
+        // Invalid format.
+        result = false;
+    }
+    else
+    {
+        // Translate the hour.
+        {
+            std::string hour_str = stime.substr(0, colon);
+            h_value = strtol(hour_str.c_str(), nullptr, 10);
+            if (h_value < 0 || h_value > 23)
+            {
+                result = false;
+            }
+        }
+
+        // Translate the minute.
+        {
+            std::string min_str = stime.substr(colon + 1);
+            m_value = strtol(min_str.c_str(), nullptr, 10);
+            if (m_value < 0 || m_value > 59)
+            {
+                result = false;
+            }
+        }
     }
 
     if (hour)
     {
-        *hour = tm.tm_hour;
+        *hour = result ? h_value : 0;
     }
     if (min)
     {
-        *min = tm.tm_min;
+        *min = result ? m_value : 0;
     }
-    return true;
+
+    return result;
 }
 
 //
 // BroadcastTimeDefs::string_to_rate_quaraters()
 //
 int16_t BroadcastTimeDefs::string_to_rate_quarters(const std::string &srate)
 {
-    const char *rate_c_str = srate.c_str();
+    // Note: The obvious way to simplify this method would be to make use of
+    //       strtof(). However, avoiding strtof() saves nearly 8K of code space
+    //       (ARMv7m, -Os).
+
+    size_t decimal = srate.find('.');
+
     char *end;
-    float rate = strtof(rate_c_str, &end);
+    long rate;
+    bool negative = false;
 
-    if (end == rate_c_str)
+    // Get the whole number portion.
     {
-        // None of the string processed.
-        return 4;
+        std::string whole_str = srate.substr(0, decimal);
+        rate = strtol(whole_str.c_str(), &end, 0);
+        if (end == whole_str.c_str())
+        {
+            // None of the string processed.
+            return 4;
+        }
+        negative = whole_str.find('-') != std::string::npos;
     }
-    if (rate < -512)
+    rate *= 4;
+
+    // Get the fractional portion
+    if (decimal != std::string::npos)
+    {
+        std::string decimal_str = srate.substr(decimal + 1);
+        decimal_str.resize(2, '0');
+        long rate_frac = strtol(decimal_str.c_str(), &end, 10);
+        if (end != decimal_str.c_str())
+        {
+            // Value 12 is chosen so that we round up when not an exact multiple
+            // of 0.25.
+            while (rate_frac > 12)
+            {
+                rate += negative ? -1 : 1;
+                rate_frac -= 25;
+            }
+        }
+    }
+
+    // Boundary checks
+    if (rate < -2048)
```

Comment:
is there a cost to using std::clamp(..) here since it also handles ensuring the value is in the expected range?

Reply:
Definitely cleaner, I didn't realize clamp was c++17. There are other casts needs in some cases so it gets it right in all cases from what I've seen as well.